Cart CRUD 구현

dnrdls7 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 장바구니 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 장바구니에 목록 추가
- JWT 을 확인하여, user의 id를 확인.
- [컨트롤러] 장바구니에 추가 시킬 상품, 수량을 입력하였는지 확인 / 미입력시 KEY ERROR 에러 응답
- [서비스] 입력된 수량이 number 인지 확인 / 유효하지 않는 입력일 시 CART_QUANTITY_IS_NOT_NUMBER 에러 응답
- [서비스] 입력된 수량이 1개 미만, 20개 초과 인지 확인 / 유효하지 않는 입력일 시 CART_QUANTITY_IS_NOT_VALID 에러 응답
- [서비스] 이미 장바구니 리스트에 있는 상품인지 확인 / 없는 경우 = 장바구니 테이블에 추가 
- [서비스] 이미 장바구니 리스트에 있는 상품인지 확인 / 2개 이상 인 경우 = 테이블에 중복 된 데이터가 있는 것임으로 500 에러코드와 CART_CHECK_LIST_OVERLAP 에러 메시지를 응답
- [서비스] 이미 장바구니 리스트에 있는 상품인지 확인 / 1개 일 경우 = 기존 수량에 +1 을 진행 / 기존 수량에 +1 했을 때 20개 초과인지 확인 / 20개 초과될 경우 CART_CAN_NOT_MORE_QUANTITY 에러 응답

2. 장바구니 목록 조회
- JWT 을 확인하여, user의 id를 확인.
- user_id로 cart table에서 해당 user_id list를 반환

3. 장바구니 목록 수정
- JWT 을 확인하여, user의 id를 확인.
- [컨트롤러] 장바구니에 수정 시킬 상품, 수량을 입력하였는지 확인 / 미입력시 key error 에러 응답
- [서비스] 장바구니 추가와 같이 수량 유효성을 검사하여, 해당 에러 응답
- [서비스] 장바구니에 추가될 때, 검사한 유효성(cart table에 데이터가 이미 있는지, 기존 수량을 체크하는 등)은 중복적 유효성을 검사하지 않음

4. 장바구니 목록 삭제
- JWT 을 확인하여, user의 id를 확인.
- [컨트롤러] 장바구니에 삭제 시킬 상품을 입력하였는지 확인 / 미입력시 key error 에러 응답
- [서비스] 장바구니 목록을 체크하여, 요청 한 유저가 해당 상품을 장바구니 목록에 가지고 있는지 확인 / 없을 시 CART_PRODUCT_IS_NOT_VALID 에러 응답

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- FE에서 검사한 유효성이라도 실제 데이터베이스에 입력되는 작업인 만큼, BE에서도 실제 비지니스에서 추구하는 유효성을 모두 재확인할 필요가 있어서, 어떤 유효성들을 가지는지 생각해보는 시간이 오래 걸렸고, 힘들었던 것 같습니다.


<br />

## :: 기타 질문 및 특이 사항
- layered 된 분기 마다 에러를 캐치하는 것과 에러 상황을 예상하여 error throw만하고 에러들은 컨트롤러 단에서 캐치 할 수 있도록 하였는데, 미들웨어 추가를 고려하였지만, 해당 효과가 정확하게 이해되지 않습니다. ( 어떠한 장점이 있는 것인지 ) 반복되는 try catch를 줄여서 해당 layered 에서 코드의 가독성이 조금 높아지는 것 말고는 장점을 생각할 수 없는 것 같습니다.

dnrdls7 committed:
[GET] cart 구현/ cartList

dnrdls7 committed:
[PUT] cart 구현/ cartUpdate

dnrdls7 committed:
[POST, DELETE] cart 구현/ cartAdd, cartDelete

dnrdls7 committed:
car
tAdd business logic 추가

dnrdls7 committed:
cart branch merge main conflict 해결

dnrdls7 committed:
routes 컨벤션 통일

dnrdls7 committed:
business logic 수정

dnrdls7 committed:
remote main pull & merge conflict 해결

dnrdls7 committed:
cart list에 product_id 추가

dnrdls7 committed:
http 요청 put -> patch로 변경

kyeongraekim commented on api/controllers/cartController.js:
```diff
@@ -0,0 +1,79 @@
+const cartService = require("../services/cartService");
+
+const cartList = async (req, res) => {
```
- 함수명은 정확한 의미가 표현된 동사, 목적어의 형태로 컨벤션이 이루어져있습니다. getAllCartList 정도로 수정해주세요~

kyeongraekim commented on api/controllers/cartController.js:
```diff
@@ -0,0 +1,79 @@
+  }
+};
+
+const cartAdd = async (req, res) => {
```
- 다른 함수명도 제 3의 개발자의 가독성을 위해서 수정해주세요. addProductInCart 정도가 가장 명확한 의미전달을 할 수 잇겟네요!

kyeongraekim commented on api/controllers/cartController.js:
```diff
@@ -0,0 +1,79 @@
+    res.status(200).json({ "data" : "cartDelete" });
+  } catch (err) {
+    console.log(err);
+    res.status(err.statusCode || 500).json({ "message" : err.message });
```
- 이미 머지된 코드 내에서는 globalErrorHandler가 존재하고 있음이 확인되는데, 그렇다면 try-catch 문에서 json 메세지를 바로 전달하면 해당 에러 핸들러의 활용성이 다소 떨어지는 형태로 보여집니다. 혹시 백엔드 팀 내부에서는 어떤 토의결과가 나왔을까요?

kyeongraekim commented on api/models/cartDao.js:
```diff
@@ -0,0 +1,76 @@
+
+const cartList = async ( user_id ) => {
+	return await appDataSource.query(`
+		SELECT 
+      products.id AS product_id,
+			name AS product_name,
+			quantity AS product_quantity,
+			price AS product_price,
+			thumbnail AS product_img,
+      user_id
+		FROM carts
```
indentation 한 번 신경 써주세요!

kyeongraekim commented on api/services/cartService.js:
```diff
@@ -0,0 +1,77 @@
+const cartUpdate = async ( user_id, product_id, quantity ) => {
+  quantity = +quantity;
+
+  if ( isNaN(quantity) ) {
+    const err = new Error('CART_QUANTITY_IS_NOT_NUMBER');
+    err.statusCode = 400;
+    throw err;
+  }
```
- 이부분은 데이터 타입에 맞게 올바르게 들어왔는지를 보는 내용이니, 비즈니스 로직을 담당하는 서비스 레이어까지 내려올 필요 없이  Controller에서 먼저 핸들링을 하면 더 좋을 것 같아 보이네요!

dnrdls7 committed:
cart 변경 사항
1. 함수명 변경
2. globalErrorHandler 적용
3. updateOptionInCartList quantity 검증 로직 service -> controller단 이동
4. 쿼리 indentation 수정

dnrdls7 committed:
cart 변경 사항
1. delete 기능 변경 / 장바구니에서 여러 상품을 체크하여 한번에 삭제할 수 있도록 변경

dnrdls7 commented:
cart 변경 사항
 1. 리뷰 반영
  - 함수명 변경
  - globalErrorHandler 적용
  - updateOptionInCartList quantity 검증 로직 service -> controller단 이동
  - 쿼리 indentation 수정
 2. 추가 기능 구현
  - 장바구니에서 여러 상품들을 선택하여 한번에 삭제할 수 있도록 delete 요청에 대한 로직 구현

kyeongraekim commented on api/controllers/cartController.js:
```diff
@@ -0,0 +1,76 @@
+});
+
+const addProductInCartList = catchAsync(async (req, res) => {
+  const { product_id, quantity } = req.body;
```
javascript naming convention에 맞게 변수명은 snake_case가 아닌 camelCase로 유지해주세요~

dnrdls7 replied:
해당 부분은 FE에서 보내주는 데이터명이라서 사용했는데, 혹시 FE랑 합의하여 바꿔야 할까요?

kyeongraekim commented on api/services/cartService.js:
```diff
@@ -0,0 +1,95 @@
+    throw err;
+  }
+
+  if ( quantity<1 || quantity>20 ) {
```
웹 개발시 필요한 상수데이터는 직접 변수명을 두고 선언한 이후 활용해주셔야 합니다. const UPPER_LIMIT, LOWER_LIMIT 등의 변수명을 활용하여서 숫자로 하드코딩이 되어있는 부분을 전체적으로 다 수정해주세요~

kyeongraekim commented on api/services/cartService.js:
```diff
@@ -0,0 +1,95 @@
+  }
+  
+  if ( Array.isArray(product_info) ) {
+    for(let i=0; i<product_info.length; i++) {
```
웹 개발의 변수명 선언 시 i와 같은 의미가 불분명한 용어 사용은 지양해주셔야 합니다. 더 올바른 의미의 변수명으로 치환시켜주세요~

kyeongraekim commented on api/models/cartDao.js:
```diff
@@ -0,0 +1,76 @@
+
+const cartListCheck = async ( user_id, product_id ) => {
+	return await appDataSource.query(`
+		SELECT 
+			name,
+			quantity,
+			price,
+			thumbnail,
+      user_id
+		FROM carts
```
indentation 한 번만 정렬 신경써 주세요~

kyeongraekim commented on api/services/cartService.js:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  const cartListCheck = await cartDao.cartListCheck( user_id, product_id );
+
+  if ( !cartListCheck || cartListCheck.length === 0 ) {
+    await cartDao.addProductInCartList( user_id, product_id, quantity );
+  } else if ( cartListCheck.length > 1 ) {
+      const err = new Error('CART_CHECK_LIST_OVERLAP');
+      err.statusCode = 500;
+      throw err;
+  } else {
+    const oldQuantity = cartListCheck[0]['quantity'];
+    if ( oldQuantity+1 > 20 ) {
+      const err = new Error('CART_CAN_NOT_MORE_QUANTITY');
```
에러 메세지가 영어 동사가 빠졌네요! ADD 정도로 올바른 문법에 맞는 에러메세지를 반환해주세요~

kyeongraekim replied:
전체 내용이 꼭 if - else if - else 로의 복잡한 분기 chaining이 필요할까요? 만일 if 단문으로 풀어낼 수 있다면 그렇게 진행하는 것이 전체 if문을 훑지 않아도 되어서 더 좋은 방향일 수 있겠네요!~

dnrdls7 committed:
cart 변경 사항
camelCase snakeCase 통일
상수데이터 변수화
indentation 수정
err 메시지 수
if else if esle -> if 통일

kyeongraekim reviewed (approved):
LGTM!~
